### PR TITLE
Blobstore Data Servlet Store picture in database

### PR DESCRIPTION
**Pull request by lilian-hunt**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+public class BlobstoreUploadUrlServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstoreService.createUploadUrl("/data");
+
+    response.setContentType("text/html");
+    response.getWriter().println(uploadUrl);
+  }
+}
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java]: This was already in https://github.com/lilian-hunt/step/pull/11/files

See https://www.michaelagreiler.com/stacked-pull-requests/

**Review by jupi3200 — COMMENTED** [review]
Would it be possible to provide a better name to pull requests.
The name should be informative and help the reader understand what functionality is added.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -51,10 +61,11 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     Map<String, Map<String, String>> comments = new HashMap<>();
     for (Entity entity : results.asIterable()) {
       String id = KeyFactory.keyToString(entity.getKey());
-      Map<String, String> commentEmail = new HashMap<>();
-      commentEmail.put("comment", (String) entity.getProperty("comment"));
-      commentEmail.put("userEmail", (String) entity.getProperty("userEmail"));
-      comments.put(id, commentEmail);
+      Map<String, String> commentAttributes = new HashMap<>();
+      commentAttributes.put("comment", (String) entity.getProperty("comment"));
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: As all attributes are related to a comment, should the this be 'text' or 'commentText'?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -77,24 +88,61 @@ private String toJSONString(Map<String, Map<String, String>> map, String name) {
...
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // Use the relative path to the image.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:145]: Please add a comment explaining why we need to apply this manipulation.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -51,10 +61,11 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
     for (Entity entity : results.asIterable()) {
       String id = KeyFactory.keyToString(entity.getKey());
-      Map<String, String> commentEmail = new HashMap<>();
-      commentEmail.put("comment", (String) entity.getProperty("comment"));
-      commentEmail.put("userEmail", (String) entity.getProperty("userEmail"));
-      comments.put(id, commentEmail);
+      Map<String, String> commentAttributes = new HashMap<>();
+      commentAttributes.put("comment", (String) entity.getProperty("comment"));
+      commentAttributes.put("userEmail", (String) entity.getProperty("userEmail"));
+      commentAttributes.put("imageUrl", (String) entity.getProperty("imageUrl"));
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:67]: If I read the code correctly, we allow client to send comments without any image.
When the client tries to retrieve the comment, they would get a json response with the string 'null' for imageURL.

Am I correct?

Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:67]: Yes that is correct

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:145]: Added a comment

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Yes this should be commentText, thanks for picking up on this.

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java (hunk 1 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java]: Ok sorry the next pull request will be to the relative branch before it .

Comment (jupi3200) [reply portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java]: OK

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (jupi3200) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:67]: That would force the client to try comparing the returned imageURL to this special value, which might be the less intuitive API.

It is ok, but perhaps a better design would not assign/send this attribute with a special value (the client can then check if the returned value has the imageURL attribute using https://www.w3schools.com/jsref/met_element_hasattribute.asp )

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -51,6 +51,7 @@ function getFeedback() {
 
     const commentListElement = document.getElementById('comment-container');
     commentListElement.innerHTML = '';
+    console.log(comments);
```
Comment (jupi3200) [inline portfolio/src/main/webapp/script.js]: Should this be removed?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (jupi3200) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:67]: How would you like to address this?

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (lilian-hunt) [reply portfolio/src/main/webapp/script.js]: Yes, has been removed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:67]: I'm not quite sure I understand what you mean by this. This method applies to the HTML dom elements, this is the map of comments and their corresponding information that I am passing to the JSON function that will eventually be read in the JS file and dynamically create the comments on the page. Could you explain what you would like to see here a little more please?

**Pull request closed without merging**